File: v2/td5-container-braitenberg/braitenberg.py
import socket
import pickle
import numpy as np


TIME_STEP = 64
MAX_SPEED = 10
RANGE = (1024 / 2)


matrice_braitenberg_droite=np.array([4,5,7,-6,-4,-2,-0.5,-0.5])
matrice_braitenberg_gauche=np.array([-2,-3,-7,7,5,4,-0.5,-0.5])


def run_client():
    vitess_moteur_droit=moteur_gauche=0
    # Create a TCP/IP socket
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    # Connect the socket to the server
    server_address = ('host.docker.internal', 8889)
    client_socket.connect(server_address)
  
    # Receive data from the server
    data = client_socket.recv(1024)
    data = pickle.loads(data)

    vitesse_moteur_gauche=np.dot(1-(np.array(data)/RANGE),matrice_braitenberg_gauche)
    vitesse_moteur_droit=np.dot(1-(np.array(data)/RANGE),matrice_braitenberg_droite)
    vitesse_moteur_gauche=np.max(vitesse_moteur_gauche)
    vitesse_moteur_droit=np.max(vitesse_moteur_droit)
    # Send data to the server
    send_data = [vitesse_moteur_gauche, vitesse_moteur_droit]
    message = pickle.dumps(send_data) 
    client_socket.sendall(message)
        
    
    # Clean up the connection
    client_socket.close()

if __name__ == '__main__':
    while True:
        run_client()

File: v2/td5-container-braitenberg/braitenberg.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h> 
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#define HOST_IP "127.0.1.1"
#define PORT_NO 8889
#define BUF_SIZE 1024

int main(int argc, char *argv[]) {
    int client_socket;
    struct sockaddr_in server_address;
    char buffer[BUF_SIZE];

    // Create a socket
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket < 0) {
        perror("Error: socket creation failed");
        exit(1);
    }

    // Configure server address
    bzero((char *) &server_address, sizeof(server_address));
    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(PORT_NO);

    // Convert the host IP from string to binary
    if (inet_pton(AF_INET, HOST_IP, &server_address.sin_addr) <= 0) {
        perror("Error: invalid host IP address");
        exit(1);
    }

    // Connect to the server
    if (connect(client_socket, (struct sockaddr *) &server_address, sizeof(server_address)) < 0) {
        perror("Error: connect to server failed");
        exit(1);
    }

    // Send data to the server
    bzero(buffer, BUF_SIZE);
    strcpy(buffer, "Hello, Server!");
    int n = write(client_socket, buffer, strlen(buffer));
    if (n < 0) {
        perror("Error: write to server failed");
        exit(1);
    }

    // Close the socket
    close(client_socket);
    return 0;
}